source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,446 @@
+  auto begin = path.begin();
+  auto read = std::next(begin);
+  auto write = std::next(begin);
+  auto end = path.end();
+  absl::string_view path_view{path};
+
+  while (read != end) {
+    if (*read == '.') {
+      char prev = *std::prev(write);
+      if (prev == '/') {
+        // attempt to read ahead 2 characters to see if we are in a "./" or "../" segment.
+        const auto dot_segment = path_view.substr(std::distance(begin, read), 3);
+        if (absl::StartsWith(dot_segment, "./") || dot_segment == ".") {
+          // This is a "/./" segment or the path is terminated by "/.", ignore it
+          size_t distance = std::min<size_t>(dot_segment.size(), 2);
+          // Advance the read iterator by 1 if the path ends with "." or 2 if the segment is "./"
+          std::advance(read, distance);
+        } else if (dot_segment == "../" || dot_segment == "..") {
+          // This is a "/../" segment or the path is terminated by "/..", navigate one segment up.
+          // Back up write 1 position to the previous slash to find the previous segment start.
+          auto new_write = findStartOfPreviousSegment(std::prev(write), begin);
+          if (new_write == begin) {
+            // This is an invalid ".." segment, most likely the full path is "/..", which attempts
+            // to go above the root.
+            return {PathNormalizationResult::Action::Reject,
+                    UhvResponseCodeDetail::get().InvalidUrl};
+          }
+
+          // Set the write position to overwrite the previous segment
+          write = new_write;
+          // Advance the read iterator by 2 if the path ends with ".." or 3 if the segment is "../"
+          size_t distance = std::min<size_t>(dot_segment.size(), 3);
+          std::advance(read, distance);
+        } else {
+          *write++ = *read++;
+        }
+      } else {
+        *write++ = *read++;
+      }
+    } else {
+      *write++ = *read++;
+    }
+  }
+
+  path.resize(std::distance(begin, write));
+  return PathNormalizationResult::success();
+}
+
+std::tuple<absl::string_view, absl::string_view>
+PathNormalizer::splitPathAndQueryParams(absl::string_view pathAndQueryParams) const {
+  // Split on the query (?) or fragment (#) delimiter, whichever one is first.
+  auto delim = pathAndQueryParams.find_first_of("?#");
+  if (delim == absl::string_view::npos) {
+    // no query/fragment component
+    return std::make_tuple(pathAndQueryParams, "");
+  }
+
+  return std::make_tuple(pathAndQueryParams.substr(0, delim), pathAndQueryParams.substr(delim));
+}
+
+std::tuple<absl::string_view, absl::string_view>
+PathNormalizer::splitAuthorityAndPath(absl::string_view uri) const {
+  // We have to make an efficient best guess for the URL form here. The URL and path will be
+  // validated further on in the UHV stack, so we don't need to be precise, just enough to split
+  // the path component and everything before it.
+  if (absl::StartsWith(uri, "//")) {
```

Comment:
have dug deeper into this.
Envoy [expects](https://github.com/envoyproxy/envoy/blob/9ae348cc43e0341ec7eadd6a4addce587afe1b27/source/common/http/conn_manager_impl.cc#L1004) the `:path` header to always be in the `origin-form` (unless it is CONNECT, in which case the path is required to be empty). Also Envoy currently rejects asterisk-form for OPTIONS requests.

The comment in the code I have linked indicates that H/1 codec is expected to parse the absolute-form target and split into the `:authority` and `:path` which happens [here](https://github.com/envoyproxy/envoy/blob/9ae348cc43e0341ec7eadd6a4addce587afe1b27/source/common/http/http1/codec_impl.cc#L1067).

Given this, UHV should expect the `:path` to be in origin form, empty for CONNECT or * for OPTIONS (this is a breaking change actually, but we'll take care of this later). And reject all other forms. In this case path normalization only happens on the origin-form and there is no need to split authority.

Reply:
I believe I've addressed this. The path normalizer now validates that the path is in origin-form, asterisk-form (`OPTIONS` method), or empty (`CONNECT` method).